### PR TITLE
virtio_serial: test ioeventfd property with virtio_serial_pci

### PR DESCRIPTION
**Pull request by nanliu-r**
ID: 1755682

Depends on [avocado-framework/avocado-vt#2376](https://github.com/avocado-framework/avocado-vt/pull/2376)

Signed-off-by: Nana Liu <nanliu@redhat.com>

### CODE REVIEW

**Comment by nanliu-r** [comment]
Test results:
Linux:
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial_ioeventfd: PASS (106.11 s)

Windows:
 (1/1) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial_ioeventfd: PASS (194.19 s)

**Comment by nanliu-r** [comment]
@vivianQizhu @yanan-fu Please help to review this together with [avocado-framework/avocado-vt#2376](https://github.com/avocado-framework/avocado-vt/pull/2376), thanks a lot.

**Comment by yanan-fu** [comment]
@nanliu-r  You can add depends on ***   in  comment 0

**Comment by nanliu-r** [comment]
> @nanliu-r You can add depends on *** in comment 0

Thanks and update.

**Comment by nanliu-r** [comment]
Test results:

 (1/3) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (1018.65 s)
 (2/3) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (139.45 s)
 (3/3) Host_RHEL.m8.u2.qcow2.virtio_blk.up.virtio_net.Guest.Win2016.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial_ioeventfd: PASS (178.18 s)

File: qemu/tests/cfg/ioeventfd.cfg [hunk 1]
```diff
@@ -1,11 +1,19 @@
 - ioeventfd: install setup image_copy unattended_install.cdrom
     type = ioeventfd
-    only virtio_blk virtio_scsi
+    only virtio_blk virtio_scsi virtio_serial
```
Comment (yanan-fu) [inline qemu/tests/cfg/ioeventfd.cfg]: We only put the options for one same `variants` together.
Move `virtio_serial` to a separated line.

Comment (yanan-fu) [reply qemu/tests/cfg/ioeventfd.cfg]: And, where is `virtio_serial` ?

File: qemu/tests/cfg/ioeventfd.cfg [hunk 2]
```diff
@@ -25,4 +33,18 @@
                 iozone_path = 'WIN_UTILS:\Iozone\iozone.exe'
                 iozone_options = '-azR -r 64k -n 1G -g 4G -M -I -i 0 -i 1 -b iozone.xls -f c:\testfile'
         - check_property:
-            compare_fd = yes
+             compare_fd = yes
```
Comment (yanan-fu) [inline qemu/tests/cfg/ioeventfd.cfg]: Wrong index

File: qemu/tests/cfg/ioeventfd.cfg [hunk 3]
```diff
@@ -25,4 +33,18 @@
...
+             serial_type_vs1 = virtserialport
+             file_transfer_serial_port = vs1
+             check_drive_format = no
+             data_transfer = yes
+             Windows:
+                 driver_name = vioser
+                 guest_script_folder = C:\
+                 clean_cmd = del /f /q
+                 tmp_dir = %TEMP%
+                 python_bin = python2.7
```
Comment (yanan-fu) [inline qemu/tests/cfg/ioeventfd.cfg]: Only python2.7 ?

File: qemu/tests/cfg/ioeventfd.cfg (hunk 1 above)
Comment (nanliu-r) [reply qemu/tests/cfg/ioeventfd.cfg]: Thanks and update, and will delete `virtio_serial`.

File: qemu/tests/cfg/ioeventfd.cfg (hunk 2 above)
Comment (nanliu-r) [reply qemu/tests/cfg/ioeventfd.cfg]: thanks and update.

File: qemu/tests/cfg/ioeventfd.cfg (hunk 3 above)
Comment (nanliu-r) [reply qemu/tests/cfg/ioeventfd.cfg]: I think it is the default version that we use inside window

**Comment by yanan-fu** [comment]
When test virtio serial ioeventfd, `only virtio_blk virtio_scsi` is not must,  it do not related with this test scenario.
suggest:
```
- ioeventfd: 
    type = ioeventfd
    ...
    variants: 
        - @block_ioeventfd:
            only virtio_blk virtio_scsi
            variants:
                - under_stress:                                                            
                    ....
                - check_property: 
                    ...
        - virtio_serial_ioeventfd: 
            check_os_type = yes
            ...  
```

File: qemu/tests/cfg/ioeventfd.cfg [hunk 4]
```diff
@@ -1,28 +1,50 @@
...
+    file_sender = both
+    transfer_timeout = 720
+    clean_cmd = rm -f
+    filesize = 10
+    host_script = serial_host_send_receive.py
+    guest_script = VirtIoChannel_guest_send_receive.py
+    variants type_dev:
+        - @block:
+            only virtio_blk virtio_scsi
+            - under_stress:
```
Comment (yanan-fu) [inline qemu/tests/cfg/ioeventfd.cfg]: Miss `variants:` above

File: qemu/tests/cfg/ioeventfd.cfg [hunk 5]
```diff
@@ -1,28 +1,50 @@
...
-                iozone_options = '-azR -r 64k -n 1G -g 4G -M -I -i 0 -i 1 -b iozone.xls -f c:\testfile'
-        - check_property:
+    guest_scripts = VirtIoChannel_guest_send_receive.py;windows_support.py
+    guest_script_folder = /var/tmp/
+    file_sender = both
+    transfer_timeout = 720
+    clean_cmd = rm -f
+    filesize = 10
+    host_script = serial_host_send_receive.py
+    guest_script = VirtIoChannel_guest_send_receive.py
```
Comment (yanan-fu) [inline qemu/tests/cfg/ioeventfd.cfg]: Please move the serial special cfg into serial variant

File: qemu/tests/ioeventfd.py [hunk 6]
```diff
@@ -41,13 +42,36 @@ def run(test, params, env):
...
+
+        :param ioeventfd: the value of parameter ioeventfd
+        :return: device id with parameter ioeventfd
+        """
+        type_dev = params.get("type_dev")
+        if type_dev == "virtio_serial":
+            params['virtio_serial_extra_params'] = ioeventfd
+            dev_id = 'virtio_serial_pci0'
+            return dev_id
         if params['drive_format'] == 'virtio':
```
Comment (yanan-fu) [inline qemu/tests/ioeventfd.py]: We do not test both serial and block device ioenventfd at the same time.
use `else` please

File: qemu/tests/cfg/ioeventfd.cfg (hunk 4 above)
Comment (nanliu-r) [reply qemu/tests/cfg/ioeventfd.cfg]: thanks and update

File: qemu/tests/ioeventfd.py (hunk 6 above)
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: The operation of `return` will separate the two tests.

**Comment by nanliu-r** [comment]
Test Pass.

 (1/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (413.02 s)
 (2/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (80.32 s)
 (3/3) Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial: PASS (116.52 s)

**Comment by nanliu-r** [comment]
@qingwangrh, please note the change of cfg and I will help to update the loop scripts once this patch is merged.

File: qemu/tests/ioeventfd.py (hunk 6 above)
Comment (yanan-fu) [reply qemu/tests/ioeventfd.py]: yes, you are right,  i missed the `return`.   but  it is dupl `return`, no need to call it for every scenario.

**Comment by qingwangrh** [comment]
@kimi1978wy @zhenyzha @peixiu Please help to review.

**Comment by peixiu** [comment]
virtio-win scsi has no this case, so ignore me here~

File: qemu/tests/ioeventfd.py (hunk 6 above)
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: Thanks and update.

File: qemu/tests/ioeventfd.py [hunk 7]
```diff
@@ -41,14 +42,37 @@ def run(test, params, env):
...
+        Set the ioeventfd options.
+
+        :param ioeventfd: the value of parameter ioeventfd
+        :return: device id with parameter ioeventfd
+        """
+        type_dev = params.get("type_dev")
+        if type_dev == "virtio_serial":
+            params['virtio_serial_extra_params'] = ioeventfd
+            dev_id = 'virtio_serial_pci0'
+            return dev_id
```
Comment (yanan-fu) [inline qemu/tests/ioeventfd.py]: no need

**Comment by qingwangrh** [comment]
python ConfigTest.py --testcase=ioeventfd --guestname=RHEL.7.8 --driveformat=virtio_blk,virtio_scsi --clone=no 
(1/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (3902.39 s)
 (2/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (145.02 s)
 (3/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial: FAIL: The "ioeventfd=off" mismatches with qtree device "virtio_serial_pci0"(false). (75.23 s)
 (4/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (4021.23 s)
 (5/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (360.68 s)
 (6/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial: FAIL: The "ioeventfd=off" mismatches with qtree device "virtio_serial_pci0"(false). (87.09 s)


Pease check the error case:
My ENV:
3.10.0-1113.el7.x86_64
qemu-kvm-rhev-2.12.0-42.el7.x86_64

File: qemu/tests/ioeventfd.py [hunk 8]
```diff
@@ -152,12 +172,15 @@ def _compare_ioeventfds():
...
-        dev_id = _set_ioeventfd_options()
+        dev_id = _set_ioeventfd_options(ioeventfd)
         error_context.context('Boot a guest with "%s".' % ioeventfd, logging.info)
         env_process.preprocess_vm(test, params, env, params["main_vm"])
         vm = env.get_vm(params["main_vm"])
         vm.verify_alive()
         session = vm.wait_for_login(timeout=timeout)
+        if os_type == 'windows' and params.get('check_os_type', ' ') != 'no':
+            session = utils_test.qemu.windrv_check_running_verifier(
+                session, vm, test, params["driver_name"])
```
[the remark this answers is not in the document]
Comment (yanan-fu) [reply qemu/tests/ioeventfd.py]: For windows guest, driver status is not important when `_check_property`. Only when do some action with that driver need ensure the driver status is ok.
how do you think about :
for case `check_property`:  no windrv_check_running_verifier, 
else, call `windrv_check_running_verifier` first for windows guest, then do other action, stress, iozone, transfer date

**Comment by nanliu-r** [comment]
> python ConfigTest.py --testcase=ioeventfd --guestname=RHEL.7.8 --driveformat=virtio_blk,virtio_scsi --clone=no
> (1/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (3902.39 s)
> (2/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (145.02 s)
> (3/6) Host_RHEL.m7.u8.qcow2.virtio_blk.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial: FAIL: The "ioeventfd=off" mismatches with qtree device "virtio_serial_pci0"(false). (75.23 s)
> (4/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.under_stress: PASS (4021.23 s)
> (5/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.check_property: PASS (360.68 s)
> (6/6) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.ioeventfd.virtio_serial: FAIL: The "ioeventfd=off" mismatches with qtree device "virtio_serial_pci0"(false). (87.09 s)
> 
> Pease check the error case:
> My ENV:
> 3.10.0-1113.el7.x86_64
> qemu-kvm-rhev-2.12.0-42.el7.x86_64

Thanks for your test, seems you miss one patch of avocado-vt: https://github.com/avocado-framework/avocado-vt/pull/2376
They need to be tested together.

File: qemu/tests/ioeventfd.py [hunk 9]
```diff
@@ -41,14 +42,36 @@ def run(test, params, env):
...
+    def _set_ioeventfd_options(ioeventfd):
+        """
+        Set the ioeventfd options.
+
+        :param ioeventfd: the value of parameter ioeventfd
+        :return: device id with parameter ioeventfd
+        """
+        type_dev = params.get("type_dev")
+        if type_dev == "virtio_serial":
+            params['virtio_serial_extra_params'] = ioeventfd
```
[the remark this answers is not in the document]
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: thanks and update.

File: qemu/tests/ioeventfd.py [hunk 10]
```diff
@@ -41,14 +42,36 @@ def run(test, params, env):
...
-        """ Set the ioeventfd options. """
-        if params['drive_format'] == 'virtio':
+    def _set_ioeventfd_options(ioeventfd):
+        """
+        Set the ioeventfd options.
+
+        :param ioeventfd: the value of parameter ioeventfd
+        :return: device id with parameter ioeventfd
+        """
+        type_dev = params.get("type_dev")
```
[the remark this answers is not in the document]
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: ok.

File: qemu/tests/ioeventfd.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: Ok, Let's separate the `transfer_data` to a single function.

File: qemu/tests/ioeventfd.py [hunk 11]
```diff
@@ -152,12 +173,15 @@ def _compare_ioeventfds():
...
     ioeventfds = (params['orig_ioeventfd'], params['new_ioeventfd'])
     for ioeventfd in ioeventfds:
-        dev_id = _set_ioeventfd_options()
+        dev_id = _set_ioeventfd_options(ioeventfd)
         error_context.context('Boot a guest with "%s".' % ioeventfd, logging.info)
         env_process.preprocess_vm(test, params, env, params["main_vm"])
         vm = env.get_vm(params["main_vm"])
         vm.verify_alive()
         session = vm.wait_for_login(timeout=timeout)
+        if os_type == 'windows' and params.get('check_os_type', ' ') != 'no':
```
[the remark this answers is not in the document]
Comment (nanliu-r) [reply qemu/tests/ioeventfd.py]: code has been updated, and I deleted this already.

**Comment by yanan-fu** [comment]
https://github.com/avocado-framework/avocado-vt/pull/2376 have merged.

**Pull request merged**